# Reference-based assembly for O1 Vibro cholerae

This is a walkthrough of a workflow to consolidate reports to assess the quality of reference-based assembly of Vibrio 
cholerae genomes.

# Contents
{:.no_toc}

1. Table of Contents
   {:toc}

## Description of workflows
The exercise will utilize workflows based on the bacpage bioinformatics pipeline.
The workflow itself is available on [Dockstore](https://dockstore.org/workflows/github.com/CholGen/bacpage/bacpage-quality-assessment:master?tab=info).
The source code for this workflow as well as bacpage is available on [GitHub](https://github.com/CholGen/bacpage/tree/master/bacpage/workflows/multiqc.wdl)
This workflow is designed for execution on the [Terra](https://terra.bio) platform, but can also be run on other cloud
platforms as well as on the command line using miniWDL.

Briefly, the workflow combines individuals quality control reports generated by the reference-based assembly 
pipelines into a single HTML report for easy viewing using [MultiQC](https://multiqc.info/).
The process combines reports from six tools:
 - `FastQC`: assesses quality of raw sequencing reads
 - `Qualimap BamQC`: assesses the quality of alignments
 - `Samtools stats`: collects statistics from alignment
 - `Samtools idxstats`: reports reference-based statistics from alignments
 - `QUAST`: assess the quality of _de novo_ assemblies
 - `BUSCO`: assesses the completeness of assemblies using universal single-copy genes

## Terra workspace setup
This tutorial assumes you have run the workflows presented in [Bacterial _de novo_ assembly and typing](workshop/theiaprok-illumina.md) 
and [Reference based assembly for O1 _Vibro cholerae_](workshop/bacpage-assemble.md).

## Analysis walkthrough
From your workspace, click on the Workflows tab on the top. This should lead to a list of analysis workflows that have
already been preloaded into your
workspace. One of them is `bacpage-quality-assessment`. Click on `bacpage-quality-assessment`.
This will lead to a workflow launch page where you will need to set the parameters and inputs before launching your analyses.
Make sure to set the following:
- The `bacpage-quality-assessment` "Version:" should be already set to `master`, but make sure it is set as such.
- "Run workflow(s) with inputs defined by data table" should be selected (not "file paths").
- "Step 1 -- Select root entity type:" should be set to `sample_set`.
- "Step 2 -- SELECT DATA" -- click on this button and a data selector box will pop up. Check box all four rows of
  the `sample` table so that we launch a job for each sample in the table. Hit the OK button
  on the lower right of the pop up box. This should return you to the workflow setup page which should now say that
  it will run on "1 selected sample_set" .
- In the inputs table, we will need to populate the following required inputs:
  - `MultiQC.bamqc_data` (required) should be set to `this.samples.bamqc_data`
  - `MultiQC.busco_reports`  should be set to `this.samples.busco_report`
  - `MultiQC.fastqc_data` should be set to `this.samples.fastqc_data`
  - `MultiQC.quast_reports` should be set to `this.samples.quast_report`
  - `MultiQC.samtools_idxstats` should be set to `this.samples.samtools_idxstats`
  - `MultiQC.samtools_stats` should be set to `this.samples.samtools_stats`
- In the outputs table, select `Use defaults` in the upper right cell.
- Click the SAVE button after you've set all the inputs. 

%Update this after seeing the Broad's set%

Click "RUN ANALYSIS" (which should be dark blue if you've filled in all inputs properly). This will take you to a job submission
status page for your newly launched workflow, showing a single row in the bottom table corresponding to the job you 
submitted.

### Wait for job completion
You will receive an email when your submission completes (along with information about whether they 
succeeded or failed). Additionally, you can also click on the JOB HISTORY tab at the top of your workspace to check 
on the status of your analyses in progress. When `bacpage-quality-assessment` is complete, you can move on to 
evaluating the results. The entry for your job under the Job History tab should have a green checkmark if the job 
was successful  

## Evaluating results
You can examine the outputs and results of each step of each job via the Job History page, however, for this workflow 
submissions, it is easier to view the saved top level outputs in the data table--in this case, the `sample_set` 
table. The `sample_set` table now has an additional output column, corresponding to the MultiQC report.   

You can click on the live links for any file element in the Terra data table and download them, or preview them in 
your browser. Click on the `multiqc.html` live link in the `multiqc_report` column of the `VibrioCholerae` row of 
the `sample_set` table. This will open a "File Details" box where you can access this file three different ways:
1. Click the blue DOWNLOAD button to download it to your computer. You can then open the downloaded file in a browser to view.
2. If you have the gcloud API installed in a command line environment, copy and paste the `gsutil cp` command to 
   your terminal, and it will download the file that way. 
3. Click on "View this file in the Google Cloud Storage Browser".
   - This opens not the file, but its parent directory in a web browser for Google Cloud buckets. This directory will 
     contain many other files, but look for the link to the file you were looking for originally (multiqc.html), and 
     click that. This leads to an "Object details" page with information and several links for this particular file. 
     Click the "Authenticated URL" link. This will open the report in your web browser.

### General Statistics table
Consult the General Statistics table at the top of the page. 
Hover-over any column header for a brief description of the statistic. 
While a lot of statistics are shown, the statistics discussed below are the most informative.
1. Under the `Total Reads` column, confirm that your samples have a suitable number of reads and that reads are 
   relatively equal between samples. A high quality genome is difficult to generate with a low number of reads 
   (generally < 1 million). Unequal reads between samples indicates that samples were not pooled equimolar when 
   making  the final library.
2. Under the `Percent Reads Mapped` column, confirm that the majority of your reads are mapped to the reference. A 
   low percentage indicates that your sample was not pure, or was not an isolate of the pathogen of interest.
3. The `Percent Coverage` column indicates the percent of the genome that is covered by at least 10 reads (the 
   minimum number of reads to call a base at a position, by default). Check that your samples have a suitable 
   percent of the genome covered. We recommend that at least 90% of the genome be covered for a sample to be 
   included in phylogenetic analysis.
4. The `Median Depth` column indicates the median number of reads which cover each position in the reference genome.  
   This metric conveys the certainty you can have in the consensus sequence. Higher median depth (>10X) indicates 
   higher trustworthiness of your consensus sequence.
5. The `Species prediction` columns shows the predicted taxonomic classification of the _de novo_ assembly. This 
   should be consulted when the percentage of mapped reads is low, to determine what species the sample likely is.

### Plots
The plots below the general statistics table can be consulted if desired, and contain supplemental information to 
the General Statistics table.
1. FastQC 
   - `Sequence Counts`: Samples should have a suitable and relatively equal number of reads. A high fraction of 
   duplicate reads is indicative of a low-diversity sequencing library which can arise due to over-amplification, or low 
   concentration/quality input DNA.
   - `Sequence Quality`: Samples should have high quality (>20 Phred Score) across the entire length of the read. 
     Inconsistencies, such as dips in the curve or rapid drop-offs, indicated problems during the sequencing run.
2. Samtools
   - `Percent Mapped`: Samples should have a high fraction of reads mapped to the reference. If this is not the case,
     then input DNA was likely not from the pathogen of interest. 
3. QUAST
   - `Assembly Statistics`: Describes the assemblies generated for each sample. This table should be consulted to 
     confirm that the assembly length generally matches the size of the genome of the pathogen that was expected 
     (for example, the _Vibrio cholerae_ genome is 4.5 million base pairs long). Shorter assembly lengths indicates that 
     there were not enough reads to construct the entire genome, while a longer assembly length might indicate that 
     the sample was not a pure colony.
   - `Number of Contigs`: High quality assemblies should have a only a few contigs that are relatively large. A large 
     number of short contigs might indicate that you did not have enough sequencing reads to completely assembly the 
     genome, or that your input DNA was too degraded. 
4. BUSCO
   - BUSCO searches your assemblies for the presence of genes that are conserved and present as single-copies in 
     most species (called BUSCOs). The percentage of these genes that are present in your assembly indicates the 
     completeness of the 
     assembly, and the number of copies of these genes indicates how pure your input DNA was (i.e. if there are two 
     copies of an expected single-copy gene, it's likely that the DNA of two species was used for sequencing). The 
     genes that are present will also suggest the taxonomic classification of your assembly.
   - High quality and pure assemblies will have almost all `Complete and single-copy BUSCOs`, incomplete assemblies 
     will have mostly `Fragmented BUSCOs` and `Missing BUSCOs`, while contaminated and un-pure assemblies will have 
     mostly `Complete and duplicated BUSCOs`